File: restocking_app/lib/Serializers.dart
import 'model/Order.dart';
import 'model/Product.dart';
import 'model/RestockingList.dart';
import 'dart:convert';

class Serializers {
  static Order orderSerializer(String responseBody){
    Map<String, dynamic> data = json.decode(responseBody);

    List<OrderItem> orderItems = new List<OrderItem>();

    for (var item in data['order_items']) {
      orderItems.add(OrderItem(
        orderItemId: item['id'],
        quantity: item['quantity'],
        processed: item['processed'],
        product: Product(
          productId: item['product']['id'],
          name: item['product']['name'],
          size: item['product']['size'],
          colour: item['product']['colour'],
          fitting: item['product']['fitting'],
          price: item['product']['price'],
          sale: item['product']['sale'],
          productType: item['product']['product_type'],
          productCode: item['product']['product_code'],
          department: item['product']['department'],
          stockQuantity: item['product']['stock_quantity'],
          floorQuantity: item['product']['floor_quantity']
        ),
      ));
    }

    Order order = Order(
      orderId: data['id'],
      orderProcessed: data['order_processed'],
      orderDelivered: data['order_delivered'],
      deliveryDate: data['delivery_date'],
      orderItems: orderItems,
    );

    return order;
  }
  
  //Not actually tested yet don't blame me if it breaks.
  static Product productSerializer(String responseBody){
    Map<String, dynamic> data = json.decode(responseBody);

    return Product(
      productId: data['id'],
      name: data['name'],
      size: data['size'],
      colour: data['colour'],
      fitting: data['fitting'],
      price: data['price'],
      sale: data['sale'],
      productType: data['product_type'],
      productCode: data['product_code'],
      department: data['department'],
      stockQuantity: data['stock_quantity'],
      floorQuantity: data['floor_quantity']
    );
  }

  static RestockingList restockingListSerializer(String responseBody){
    Map<String, dynamic> data = json.decode(responseBody);

    List<RestockingListItem> restockingListItems = new List<RestockingListItem>();

    for (var item in data['restocking_items']) {
      restockingListItems.add(RestockingListItem(
        restockingListItemId: item['id'],
        quantity: item['quantity'],
        processed: item['processed'],
        product: Product(
          productId: item['product']['id'],
          name: item['product']['name'],
          size: item['product']['size'],
          colour: item['product']['colour'],
          fitting: item['product']['fitting'],
          price: item['product']['price'],
          sale: item['product']['sale'],
          productType: item['product']['product_type'],
          productCode: item['product']['product_code'],
          department: item['product']['department'],
          stockQuantity: item['product']['stock_quantity'],
          floorQuantity: item['product']['floor_quantity']
        )
      ));
    }

    return RestockingList(
      restockingListId: data['id'],
      date: data['date'],
      time: data['time'],
      restockingListItems: restockingListItems
    );
  }

  static String deserialiseOrderItemUpdate(int itemId, int processed){ 
    var jsonMap = {
      "order_items": [
        {
          "id": itemId,
          "processed":processed,
        }
      ]
    };
    return jsonEncode(jsonMap);
  }

  static String deserialiseOrderItemCreate(int productId, int orderId){
    var jsonMap = {
      "order": orderId,
      "quantity" : 1,
      "processed" : 1,
      "product" : {
        "id" : productId
      }
    };
    return jsonEncode(jsonMap);
  }

  static OrderItem orderItemSerializer(int productId){
    return OrderItem(
      product: Product(productId: productId),
      quantity: 1,
      processed: 1
    );
  }

  static String deserialiseRestockingListItemUpdate(int itemId, int processed, int productId, int restockingListId){ 
    var jsonMap = {
      "restocking_items":[
        {
          "id":itemId,
          "processed":processed,
          "product":{
            "id":productId
          },
          "restocking_list":restockingListId
        }
      ]
    };
    return jsonEncode(jsonMap);
  }

  static String deserialiseRestockingListItemCreate(int productId, int restockingListId){
    var jsonMap = {
      "restocking_list": restockingListId,
      "quantity" : 1,
      "processed" : 1,
      "product" : {
        "id" : productId
      }
    };
    return jsonEncode(jsonMap);
  }

  static RestockingListItem restockingListItemSerializer(int productId){
    return RestockingListItem(
      product: Product(productId: productId),
      quantity: 1,
      processed: 1
    );
  }

  static String deserialiseProductUpdate(int productId, int stockQuantity, int floorQuantity){
    var jsonMap = {
      "id": productId,
      "stock_quantity": stockQuantity,
      "floor_quantity": floorQuantity 
    };
    return jsonEncode(jsonMap);
  }

  static RestockingList productSerializerRecommend(String responseBody){
    Map<String, dynamic> data = json.decode(responseBody);

    List<RestockingListItem> restockingListItems = new List<RestockingListItem>(); 

    for (var item in data['recommendations']) {
      restockingListItems.add(
        RestockingListItem(
          product: Product(
            productId: item['id'],
            name: item['name'],
            size: double.parse(item['size']),
            colour: item['colour'],
            fitting: item['fitting'],
            price: item['price'],
            sale: item['sale'],
            productType: item['product_type'],
            productCode: item['product_code'],
            department: item['department'],
            floorQuantity: item['floor_quantity'],
            stockQuantity: item['stock_quantity']
          )
        )
      );
    }
    print(restockingListItems[0].product);
    return RestockingList(restockingListItems:restockingListItems);
  }
}

File: django/dsp/restocking/serializers.py
from rest_framework import serializers
from .models import Product, Order, OrderItem, RestockingList, RestockingListItem

class ProductSerializer(serializers.ModelSerializer):
    """Serializer to map a model into JSON format"""

    id=serializers.IntegerField(required=False)

    class Meta:
        model = Product
        fields = ('id', 'name', 'size', 'colour', 'fitting', 'price', 'sale', 'product_type', 'product_code', 'department', 'stock_quantity', 'floor_quantity')
        read_only_fields = ('name', 'size', 'colour', 'fitting', 'price', 'sale', 'product_type', 'product_code', 'department')

class OrderItemSerializer(serializers.ModelSerializer):
    product = ProductSerializer()

    id = serializers.IntegerField(required=False)

    class Meta:
        model = OrderItem
        fields = ('id', 'quantity', 'product', 'processed', 'order', 'added')
        read_only_fields = ['id']


class OrderSerializer(serializers.ModelSerializer):
    """Serializer to map a model into JSON Format for Orders"""
    order_items = OrderItemSerializer(many=True)

    class Meta:
        model = Order
        fields = ('id', 'order_processed', 'order_delivered', 'delivery_date', 'delivery_processed', 'order_items')
        read_only_fields = ('id', 'order_processed', 'order_delivered', 'delivery_date', 'delivery_processed')

    def update(self, instance, validated_data):
        print(validated_data)
        items = validated_data.get('order_items')
        for item in items:
            item_id = item.get('id', None)
            order_item = OrderItem.objects.get(id__exact=item_id)
            order_item.processed = item.get('processed', order_item.processed)
            order_item.quantity = item.get('quantity', order_item.quantity)
            order_item.save()

        return instance

class RestockingListItemSerializer(serializers.ModelSerializer):
    """Serializer to map a model into JSON format for restocking list items"""
    product = ProductSerializer()

    id = serializers.IntegerField(required=False)

    class Meta:
        model = RestockingListItem
        fields = ('id', 'quantity', 'product', 'processed', 'restocking_list', 'added')
        read_only_fields = ['id']

class RestockingListSerializer(serializers.ModelSerializer):
    """Serializer to map a model into JSON format for restocking lists"""
    restocking_items = RestockingListItemSerializer(many=True)

    class Meta:
        model = RestockingList
        fields = ('id', 'date', 'time', 'restocking_items', 'datetime')
        read_only_fields = ('id', 'date', 'time', 'datetime')

    def update(self, instance, validated_data):
        restocking_list_items = validated_data.get('restocking_items')
        for item in restocking_list_items:
            item_id = item.get('id', None)
            restocking_list_item = RestockingListItem.objects.get(id__exact=item_id)
            restocking_list_item.processed = item.get('processed', restocking_list_item.processed)
            restocking_list_item.quantity = item.get('quantity', restocking_list_item.quantity)
            restocking_list_item.save()

        return instance


def serialize_recommendation(candidates):
    """
    Serialises our recommendations into a list of products.
    """
    dictionary = {}
    dictionary['recommendations'] = []
    for index, item in enumerate(list(candidates)):
        dictionary['recommendations'].append({})
        dictionary['recommendations'][index]['id'] = item.id
        dictionary['recommendations'][index]['name'] = item.name
        dictionary['recommendations'][index]['size'] = str(item.size)
        dictionary['recommendations'][index]['colour'] = item.colour
        dictionary['recommendations'][index]['fitting'] = item.fitting
        dictionary['recommendations'][index]['price'] = str(item.price)
        dictionary['recommendations'][index]['sale'] = item.sale
        dictionary['recommendations'][index]['product_type'] = item.product_type
        dictionary['recommendations'][index]['product_code'] = item.product_code
        dictionary['recommendations'][index]['department'] = item.department
        dictionary['recommendations'][index]['floor_quantity'] = item.floor_quantity
        dictionary['recommendations'][index]['stock_quantity'] = item.stock_quantity

    return dictionary